src/sentry/sentry_metrics/querying/data/query.py

```diff
@@ -0,0 +1,86 @@
+import re
+from collections.abc import Mapping
+from dataclasses import dataclass
+from typing import Any, cast
+
+from sentry.sentry_metrics.querying.data.execution import QueryResult
+from sentry.sentry_metrics.querying.data.transformation.base import (
+    QueryResultsTransformer,
+    QueryTransformerResult,
+)
+from sentry.sentry_metrics.querying.errors import InvalidMetricsQueryError
+from sentry.sentry_metrics.querying.types import QueryOrder
+
+
+class MQLQuery:
+    """
+    Represents an MQL query that can be run against Snuba.
+
+    Example:
+    # Defining a simple query.
+    query = MQLQuery("avg(d:transactions/duration@millisecond)", order=QueryOrder.ASC, limit=10)
+
+    # Defining more complex queries that depend on each other.
+    query_1 = MQLQuery("avg(d:transactions/duration@millisecond)")
+    query_2 = MQLQuery("sum(d:transactions/duration@millisecond)")
+    query = MQLQuery("$query_1 / $query_2", order=QueryOrder.ASC, query_1=query_1, query_2=query_2)
+    """
+
+    def __init__(
+        self, mql: str, order: QueryOrder | None = None, limit: int | None = None, **sub_queries
+    ):
+        self.mql = mql
+        self.order = order
+        self.limit = limit
+        self.sub_queries = self._validate_sub_queries(sub_queries)
+
+    @staticmethod
+    def _validate_sub_queries(sub_queries: Mapping[str, Any]) -> Mapping[str, "MQLQuery"]:
+        for name, query in sub_queries.items():
+            if not isinstance(query, MQLQuery):
+                raise InvalidMetricsQueryError("A subquery must be an instance of 'MQLQuery'")
+
+        return cast(Mapping[str, MQLQuery], sub_queries)
+
+    def compile(self) -> "MQLQuery":
```

Comment:
nit: can we find a better name for this?

Reply:
I find that `compile` reflects what it is doing since it's assembling a final query from sub-queries. If you have any other proposals I am up to discuss them.

Reply:
`replace_formula` since that is exactly what it is doing? 😅